# Joining (Merging) DataFrames

Using the [MovieLens 100k data](http://grouplens.org/datasets/movielens/), let's create two DataFrames:

- **movies**: shows information about movies, namely a unique **movie_id** and its **title**
- **ratings**: shows the **rating** that a particular **user_id** gave to a particular **movie_id** at a particular **timestamp**

### Movies

In [2]:
import pandas as pd
movie_url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.item'
movie_cols = ['movie_id', 'title']
movies = pd.read_table(movie_url, sep='|', header=None, names=movie_cols,
                       usecols=[0, 1], encoding="ISO-8859-1")
movies.head()


movie_id              title
0         1   Toy Story (1995)
1         2   GoldenEye (1995)
2         3  Four Rooms (1995)
3         4  Get Shorty (1995)
4         5     Copycat (1995)

In [3]:
#get the size of movies dataframe (number of rows )
len(movies)

1682

### Ratings

In [4]:
rating_url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.data'
rating_cols = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_table(rating_url, sep='\t', header=None, names=rating_cols)
ratings.head()

user_id  movie_id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [5]:
len(ratings)

100000

Let's pretend that you want to examine the ratings DataFrame, but you want to know the **title** of each movie rather than its **movie_id**. The best way to accomplish this objective is by "joining" (or "merging") the DataFrames using the Pandas `merge` function:

In [6]:
movie_ratings = pd.merge(movies, ratings)
movie_ratings.head()
# movie_ratings = pd.merge(movies, ratings,
#                          left_on='movie_id',
#                          right_on='Movie_Id')


movie_id             title  user_id  rating  timestamp
0         1  Toy Story (1995)      308       4  887736532
1         1  Toy Story (1995)      287       5  875334088
2         1  Toy Story (1995)      148       4  877019411
3         1  Toy Story (1995)      280       4  891700426
4         1  Toy Story (1995)       66       3  883601324

In [7]:
pd.merge(ratings, movies)

user_id  movie_id  rating  timestamp  \
0          196       242       3  881250949   
1           63       242       3  875747190   
2          226       242       5  883888671   
3          154       242       3  879138235   
4          306       242       5  876503793   
5          296       242       4  884196057   
6           34       242       5  888601628   
7          271       242       4  885844495   
8          201       242       4  884110598   
9          209       242       4  883589606   
10          35       242       2  875459166   
11         354       242       5  891180399   
12         199       242       5  883782485   
13         113       242       2  875075887   
14           1       242       5  889751633   
15         173       242       5  877556626   
16         360       242       4  880353616   
17         234       242       4  891033261   
18          14       242       4  876964570   
19         309       242       4  877370319   
20         331       242       4  877196089   
21          21       242       3  874951617   
22         111       242       4  891679901   
23         439       242       5  882892424   
24         355       242       4  879486529   
25         204       242       5  892388935   
26         145       242       5  875269755   
27          30       242       5  885941156   
28         463       242       2  889935629   
29         144       242       4  888103444   
...        ...       ...     ...        ...   
99970      405      1564       1  885546288   
99971      456      1604       4  881372849   
99972      405      1574       1  885546529   
99973      883      1448       5  891695570   
99974      655      1448       3  887523224   
99975      234      1448       3  892335187   
99976      787      1671       1  888980193   
99977      587      1625       4  892871732   
99978      782      1665       2  891500194   
99979      181      1325       1  878962816   
99980      782      1670       3  891497793   
99981      181      1341       1  878962169   
99982      181      1352       1  878962240   
99983      655      1636       4  887473570   
99984      896      1681       3  887160722   
99985      655      1638       3  887488947   
99986      782      1667       3  891500110   
99987      851      1675       3  884222085   
99988      234      1460       3  892335460   
99989      532      1594       4  893115576   
99990      416      1594       5  893212484   
99991      648      1626       1  884795447   
99992      655      1645       4  892871225   
99993      747      1659       1  888733313   
99994      916      1682       3  880845755   
99995      840      1674       4  891211682   
99996      655      1640       3  888474646   
99997      655      1637       3  888984255   
99998      655      1630       3  887428735   
99999      655      1641       3  887427810   

                                                   title  
0                                           Kolya (1996)  
1                                           Kolya (1996)  
2                                           Kolya (1996)  
3                                           Kolya (1996)  
4                                           Kolya (1996)  
5                                           Kolya (1996)  
6                                           Kolya (1996)  
7                                           Kolya (1996)  
8                                           Kolya (1996)  
9                                           Kolya (1996)  
10                                          Kolya (1996)  
11                                          Kolya (1996)  
12                                          Kolya (1996)  
13                                          Kolya (1996)  
14                                          Kolya (1996)  
15                                          Kolya (1996)  
16                                          Kolya (1996)  
17                            

In [8]:
len(movie_ratings)


100000

In [9]:
ratings1 = movie_ratings.copy(deep=True)
ratings1 = ratings1[1:]
ratings1[0:10]


movie_id             title  user_id  rating  timestamp
1          1  Toy Story (1995)      287       5  875334088
2          1  Toy Story (1995)      148       4  877019411
3          1  Toy Story (1995)      280       4  891700426
4          1  Toy Story (1995)       66       3  883601324
5          1  Toy Story (1995)        5       4  875635748
6          1  Toy Story (1995)      109       4  880563619
7          1  Toy Story (1995)      181       3  878962392
8          1  Toy Story (1995)       95       5  879197329
9          1  Toy Story (1995)      268       3  875742341
10         1  Toy Story (1995)      189       5  893264174

In [10]:
ratings2 = movie_ratings.copy(deep=True)
ratings2[0:10]

#ratings[:-1]



movie_id             title  user_id  rating  timestamp
0         1  Toy Story (1995)      308       4  887736532
1         1  Toy Story (1995)      287       5  875334088
2         1  Toy Story (1995)      148       4  877019411
3         1  Toy Story (1995)      280       4  891700426
4         1  Toy Story (1995)       66       3  883601324
5         1  Toy Story (1995)        5       4  875635748
6         1  Toy Story (1995)      109       4  880563619
7         1  Toy Story (1995)      181       3  878962392
8         1  Toy Story (1995)       95       5  879197329
9         1  Toy Story (1995)      268       3  875742341

In [11]:
diff = ratings2['rating'] - ratings1['rating']

print(diff.head())

0    NaN
1    0.0
2    0.0
3    0.0
4    0.0
Name: rating, dtype: float64


Here's what just happened:

- Pandas noticed that movies and ratings had one column in common, namely **movie_id**. This is the "key" on which the DataFrames will be joined.
- The first **movie_id** in movies is 1. Thus, Pandas looked through every row in the ratings DataFrame, searching for a movie_id of 1. Every time it found such a row, it recorded the **user_id**, **rating**, and **timestamp** listed in that row. In this case, it found 452 matching rows.
- The second **movie_id** in movies is 2. Again, Pandas did a search of ratings and found 131 matching rows.
- This process was repeated for all of the remaining rows in movies.

At the end of the process, the movie_ratings DataFrame is created, which contains the two columns from movies (**movie_id** and **title**) and the three other colums from ratings (**user_id**, **rating**, and **timestamp**).

- **movie_id** 1 and its **title** are listed 452 times, next to the **user_id**, **rating**, and **timestamp** for each of the 452 matching ratings.
- **movie_id** 2 and its **title** are listed 131 times, next to the **user_id**, **rating**, and **timestamp** for each of the 131 matching ratings.
- And so on, for every movie in the dataset.

In [12]:
print( movies.shape)
print(ratings.shape)
print(movie_ratings.shape)

(1682, 2)
(100000, 4)
(100000, 5)


Notice the shapes of the three DataFrames:

- There are 1682 rows in the movies DataFrame.
- There are 100000 rows in the ratings DataFrame.
- The `merge` function resulted in a movie_ratings DataFrame with 100000 rows, because every row from ratings matched a row from movies.
- The movie_ratings DataFrame has 5 columns, namely the 2 columns from movies, plus the 4 columns from ratings, minus the 1 column in common.

By default, the `merge` function joins the DataFrames using all column names that are in common (**movie_id**, in this case). The [documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) explains how you can override this behavior.

## Four Types of Joins

There are actually four types of joins supported by the Pandas `merge` function. Here's how they are described by the documentation:

- **inner:** use intersection of keys from both frames (SQL: inner join)
- **outer:** use union of keys from both frames (SQL: full outer join)
- **left:** use only keys from left frame (SQL: left outer join)
- **right:** use only keys from right frame (SQL: right outer join)

The default is the "inner join", which was used when creating the movie_ratings DataFrame.

It's easiest to understand the different types by looking at some simple examples:

### Example DataFrames A and B

In [13]:
A = pd.DataFrame({'color': ['green', 'yellow', 'red'], 'num':[1, 2, 3]})
A

color  num
0   green    1
1  yellow    2
2     red    3

In [14]:
B = pd.DataFrame({'color': ['green', 'yellow', 'pink'], 'size':['S', 'M', 'L']})
B

color size
0   green    S
1  yellow    M
2    pink    L

### Inner join

Only include observations found in both A and B:

In [15]:
pd.merge(A, B, how='inner')

color  num size
0   green    1    S
1  yellow    2    M

### Outer join

Include observations found in either A or B:

In [16]:
pd.merge(A, B, how='outer')

color  num size
0   green  1.0    S
1  yellow  2.0    M
2     red  3.0  NaN
3    pink  NaN    L

### Left join

Include all observations found in A:

In [17]:
pd.merge(A, B, how='left')

color  num size
0   green    1    S
1  yellow    2    M
2     red    3  NaN

### Right join

Include all observations found in B:

In [18]:
pd.merge(A, B, how='right')

color  num size
0   green  1.0    S
1  yellow  2.0    M
2    pink  NaN    L

# Extra Fun

## get top 5 movies with:
- best avg ratings
- most number of ratings


In [19]:
movie_ratings.groupby('title')['rating'].mean().sort_values(ascending=False)

title
Marlene Dietrich: Shadow and Light (1996)                 5.000000
Prefontaine (1997)                                        5.000000
Santa with Muscles (1996)                                 5.000000
Star Kid (1997)                                           5.000000
Someone Else's America (1995)                             5.000000
Entertaining Angels: The Dorothy Day Story (1996)         5.000000
Saint of Fort Washington, The (1993)                      5.000000
Great Day in Harlem, A (1994)                             5.000000
They Made Me a Criminal (1939)                            5.000000
Aiqing wansui (1994)                                      5.000000
Pather Panchali (1955)                                    4.625000
Anna (1996)                                               4.500000
Everest (1998)                                            4.500000
Maya Lin: A Strong Clear Vision (1994)                    4.500000
Some Mother's Son (1996)                                

In [20]:
movie_ratings.groupby('title')['rating'].count().sort_values(ascending=False)

title
Star Wars (1977)                                     583
Contact (1997)                                       509
Fargo (1996)                                         508
Return of the Jedi (1983)                            507
Liar Liar (1997)                                     485
English Patient, The (1996)                          481
Scream (1996)                                        478
Toy Story (1995)                                     452
Air Force One (1997)                                 431
Independence Day (ID4) (1996)                        429
Raiders of the Lost Ark (1981)                       420
Godfather, The (1972)                                413
Pulp Fiction (1994)                                  394
Twelve Monkeys (1995)                                392
Silence of the Lambs, The (1991)                     390
Jerry Maguire (1996)                                 384
Chasing Amy (1997)                                   379
Rock, The (1996)         

In [21]:
movie_ratings.groupby('title')['rating'].describe()

count      mean       std  \
title                                                                           
'Til There Was You (1997)                             9.0  2.333333  1.000000   
1-900 (1994)                                          5.0  2.600000  1.516575   
101 Dalmatians (1996)                               109.0  2.908257  1.076184   
12 Angry Men (1957)                                 125.0  4.344000  0.719588   
187 (1997)                                           41.0  3.024390  1.172344   
2 Days in the Valley (1996)                          93.0  3.225806  0.957000   
20,000 Leagues Under the Sea (1954)                  72.0  3.500000  0.731581   
2001: A Space Odyssey (1968)                        259.0  3.969112  1.026307   
3 Ninjas: High Noon At Mega Mountain (1998)           5.0  1.000000  0.000000   
39 Steps, The (1935)                                 59.0  4.050847  0.859505   
8 1/2 (1963)                                         38.0  3.815789  1.182194   
8 Heads in a Duffel Bag (1997)                        4.0  3.250000  1.500000   
8 Seconds (1994)                                      4.0  3.750000  1.258306   
A Chef in Love (1996)                                 8.0  4.125000  0.834523   
Above the Rim (1994)                                  5.0  3.000000  0.707107   
Absolute Power (1997)                               127.0  3.370079  0.949546   
Abyss, The (1989)                                   151.0  3.589404  0.911201   
Ace Ventura: Pet Detective (1994)                   103.0  3.048544  1.215735   
Ace Ventura: When Nature Calls (1995)                37.0  2.675676  1.106899   
Across the Sea of Time (1995)                         4.0  2.750000  1.258306   
Addams Family Values (1993)                          87.0  2.816092  0.946582   
Addicted to Love (1997)                              54.0  3.166667  1.004706   
Addiction, The (1995)                                11.0  2.181818  1.328020   
Adventures of Pinocchio, The (1996)                  39.0  3.051282  1.145902   
Adventures of Priscilla, Queen of the Desert, T...  111.0  3.594595  1.030255   
Adventures of Robin Hood, The (1938)                 67.0  3.791045  0.844542   
Affair to Remember, An (1957)                        26.0  4.192308  0.895287   
African Queen, The (1951)                           152.0  4.184211  0.758362   
Afterglow (1997)                                     18.0  3.111111  1.022620   
Age of Innocence, The (1993)                         65.0  3.384615  1.010759   
...                                                   ...       ...       ...   
Window to Paris (1994)                                1.0  4.000000       NaN   
Wings of Courage (1995)                               1.0  4.000000       NaN   
Wings of Desire (1987)                               57.0  4.000000  1.118034   
Wings of the Dove, The (1997)                        75.0  3.680000  0.902848   
Winnie the Pooh and the Blustery Day (1968)          75.0  3.800000  1.090500   
Winter Guest, The (1997)                              9.0  3.444444  1.424001   
Wishmaster (1997)                                    27.0  2.444444  1.395965   
With Honors (1994)                                   46.0  3.065217  1.143137   
Withnail and I (1987)                                13.0  3.230769  1.300887   
Witness (1985)                                        1.0  4.000000       NaN   
Wizard of Oz, The (1939)                            246.0  4.077236  0.959448   
Wolf (1994)                                          67.0  2.701493  0.953700   
Woman in Question, The (1950)                         1.0  1.000000       NaN   
Women, The (1939)                                    15.0  3.666667  1.175139   
Wonderful, Horrible Life of Leni Riefenstahl, T...   10.0  4.000000  0.666667   
Wonderland (1997)                                    10.0  3.200000  1.475730   
Wooden Man's Bride, The (Wu Kui) (1994)               3.0  2.666667  1.527525   
Wo

In [22]:
movie_ratings.groupby('title')['rating'].describe().\
    sort_values(by=['count', 'mean'], ascending=[False, False])[['mean', 'count']]

mean  count
title                                                             
Star Wars (1977)                                   4.358491  583.0
Contact (1997)                                     3.803536  509.0
Fargo (1996)                                       4.155512  508.0
Return of the Jedi (1983)                          4.007890  507.0
Liar Liar (1997)                                   3.156701  485.0
English Patient, The (1996)                        3.656965  481.0
Scream (1996)                                      3.441423  478.0
Toy Story (1995)                                   3.878319  452.0
Air Force One (1997)                               3.631090  431.0
Independence Day (ID4) (1996)                      3.438228  429.0
Raiders of the Lost Ark (1981)                     4.252381  420.0
Godfather, The (1972)                              4.283293  413.0
Pulp Fiction (1994)                                4.060914  394.0
Twelve Monkeys (1995)                              3.798469  392.0
Silence of the Lambs, The (1991)                   4.289744  390.0
Jerry Maguire (1996)                               3.710938  384.0
Chasing Amy (1997)                                 3.839050  379.0
Rock, The (1996)                                   3.693122  378.0
Empire Strikes Back, The (1980)                    4.204360  367.0
Star Trek: First Contact (1996)                    3.660274  365.0
Titanic (1997)                                     4.245714  350.0
Back to the Future (1985)                          3.834286  350.0
Mission: Impossible (1996)                         3.313953  344.0
Fugitive, The (1993)                               4.044643  336.0
Indiana Jones and the Last Crusade (1989)          3.930514  331.0
Willy Wonka and the Chocolate Factory (1971)       3.631902  326.0
Princess Bride, The (1987)                         4.172840  324.0
Forrest Gump (1994)                                3.853583  321.0
Monty Python and the Holy Grail (1974)             4.066456  316.0
Saint, The (1997)                                  3.123418  316.0
...                                                     ...    ...
Low Life, The (1994)                               1.000000    1.0
Man from Down Under, The (1943)                    1.000000    1.0
Mat' i syn (1997)                                  1.000000    1.0
Mille bolle blu (1993)                             1.000000    1.0
Modern Affair, A (1995)                            1.000000    1.0
Mostro, Il (1994)                                  1.000000    1.0
New Age, The (1994)                                1.000000    1.0
Nobody Loves Me (Keiner liebt mich) (1994)         1.000000    1.0
Office Killer (1997)                               1.000000    1.0
Pharaoh's Army (1995)                              1.000000    1.0
Police Story 4: Project S (Chao ji ji hua) (1993)  1.000000    1.0
Power 98 (1995)                                    1.000000    1.0
Promise, The (Versprechen, Das) (1994)             1.000000    1.0
Quartier Mozart (1992)                             1.000000    1.0
Shadow of Angels (Schatten der Engel) (1976)       1.000000    1.0
Shadows (Cienie) (1988)                            1.000000    1.0
Stefano Quantestorie (1993)                        1.000000    1.0
Symphonie pastorale, La (1946)                     1.000000    1.0
T-Men (1947)                                       1.000000    1.0
Terror in a Texas Town (1958)                      1.000000    1.0
The Courtyard (1995)                               1.000000    1.0
Tigrero: A Film That Was Never Made (1994)         1.000000    1.0
To Cross the Rubicon (1991)                        1.000000    1.0
Touki Bouki (Journey of the Hyena) (1973)          1.000000    1.0
Vermont Is For Lovers (1992)                       1.000000    1.0
Very Natural Thing, A (1974)                       1.000000    1.0
Vie est belle, La (Life is Rosey) (1987)           1.000000    1.0
Wend Kuuni (God's Gift) (1982)     

In [29]:
movie_ratings_avg = \
    movie_ratings.groupby('title')['rating'].mean().sort_values(ascending=False)
movie_ratings_avg = movie_ratings_avg.reset_index()
movie_ratings_avg.head()

title  rating
0  Marlene Dietrich: Shadow and Light (1996)      5.0
1                          Prefontaine (1997)     5.0
2                   Santa with Muscles (1996)     5.0
3                             Star Kid (1997)     5.0
4               Someone Else's America (1995)     5.0

In [33]:
movie_ratings_count = \
    movie_ratings.groupby('title')['rating'].count().sort_values(ascending=False)

In [34]:
movie_ratings_count = movie_ratings_count.reset_index()
movie_ratings_count.columns = ['title', 'count']
movie_ratings_count.head()

title  count
0           Star Wars (1977)    583
1             Contact (1997)    509
2               Fargo (1996)    508
3  Return of the Jedi (1983)    507
4           Liar Liar (1997)    485

In [35]:
pd.merge(movie_ratings_avg, movie_ratings_count)

title    rating  count
0            Marlene Dietrich: Shadow and Light (1996)   5.000000      1
1                                    Prefontaine (1997)  5.000000      3
2                             Santa with Muscles (1996)  5.000000      2
3                                       Star Kid (1997)  5.000000      3
4                         Someone Else's America (1995)  5.000000      1
5     Entertaining Angels: The Dorothy Day Story (1996)  5.000000      1
6                  Saint of Fort Washington, The (1993)  5.000000      2
7                         Great Day in Harlem, A (1994)  5.000000      1
8                        They Made Me a Criminal (1939)  5.000000      1
9                                  Aiqing wansui (1994)  5.000000      1
10                               Pather Panchali (1955)  4.625000      8
11                                          Anna (1996)  4.500000      2
12                                       Everest (1998)  4.500000      2
13               Maya Lin: A Strong Clear Vision (1994)  4.500000      4
14                             Some Mother's Son (1996)  4.500000      2
15                                Close Shave, A (1995)  4.491071    112
16                              Schindler's List (1993)  4.466443    298
17                           Wrong Trousers, The (1993)  4.466102    118
18                                    Casablanca (1942)  4.456790    243
19    Wallace & Gromit: The Best of Aardman Animatio...  4.447761     67
20                     Shawshank Redemption, The (1994)  4.445230    283
21                                   Rear Window (1954)  4.387560    209
22                           Usual Suspects, The (1995)  4.385768    267
23                                     Star Wars (1977)  4.358491    583
24                                  12 Angry Men (1957)  4.344000    125
25                                Third Man, The (1949)  4.333333     72
26                      Letter From Death Row, A (1998)  4.333333      3
27                  Bitter Sugar (Azucar Amargo) (1996)  4.333333      3
28                                  Citizen Kane (1941)  4.292929    198
29              Some Folks Call It a Sling Blade (1993)  4.292683     41
...                                                 ...       ...    ...
1634                     Symphonie pastorale, La (1946)  1.000000      1
1635                                 Baton Rouge (1988)  1.000000      1
1636                              Boys in Venice (1996)  1.000000      2
1637         Tigrero: A Film That Was Never Made (1994)  1.000000      1
1638                               Office Killer (1997)  1.000000      1
1639                                       T-Men (1947)  1.000000      1
1640                      Terror in a Texas Town (1958)  1.000000      1
1641         Nobody Loves Me (Keiner liebt mich) (1994)  1.000000      1
1642                               The Courtyard (1995)  1.000000      1
1643                            King of New York (1990)  1.000000      1
1644                                    Liebelei (1933)  1.000000      1
1645                        Stefano Quantestorie (1993)  1.000000      1
1646                          Hostile Intentions (1994)  1.000000      1
1647                              Pharaoh's Army (1995)  1.000000      1
1648                                New Age, The (1994)  1.000000      1
1649                     Hungarian Fairy Tale, A (1987)  1.000000      1
1650  Death in the Garden (Mort en ce jardin, La) (1...  1.000000      1
1651             Promise, The (Versprechen, Das) (1994)  1.000000      1
1652      I, Worst of All (Yo, la peor de todas) (1990)  1.000000      1
1653                                    Power 98 (1995)  1.000000      1
1654  Police Story 4: Project S (Chao ji ji hua) (1993)  1.000000      1
1655                               Low Life, The (1994)  1.000000      1
1656                            Somebody to Love (1994)  1.000000      2
1657               Invitation, The 

In [37]:
movie_ratings.groupby('title')['rating'].mean()

title
'Til There Was You (1997)                                   2.333333
1-900 (1994)                                                2.600000
101 Dalmatians (1996)                                       2.908257
12 Angry Men (1957)                                         4.344000
187 (1997)                                                  3.024390
2 Days in the Valley (1996)                                 3.225806
20,000 Leagues Under the Sea (1954)                         3.500000
2001: A Space Odyssey (1968)                                3.969112
3 Ninjas: High Noon At Mega Mountain (1998)                 1.000000
39 Steps, The (1935)                                        4.050847
8 1/2 (1963)                                                3.815789
8 Heads in a Duffel Bag (1997)                              3.250000
8 Seconds (1994)                                            3.750000
A Chef in Love (1996)                                       4.125000
Above the Rim (1994)        